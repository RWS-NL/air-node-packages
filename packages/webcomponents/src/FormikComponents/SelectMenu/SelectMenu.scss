@import '../../styles/vars';

.select {
  border-radius: 0 !important;
  border-width: 1px;
  height: 2.5rem;
  min-height: 2.5rem;

  :focus {
    background-color: transparent;
  }

  :focus-within {
    background-color: transparent;
  }

  :active {
    background-color: transparent;
  }

  &Entry {
    padding-top: 0;
  }
}

.input {
  color: $textBlack;
  font-size: 0.9rem;
}

.inputLabel {
  color: $grey75;
  font-size: 1rem;
  font-style: italic;
  transform: translate(1rem, 0.8rem);

  &Shrink {
    color: $textBlack;
  }
}

.error {
  &Select > fieldset {
    border-color: $red100 !important;
  }

  &Select:hover > fieldset {
    border-width: 2px;
  }

  &Label {
    color: $red100;
    position: absolute;
    top: 2.5rem;
  }
}

.menuItem {
  font-size: 0.9rem;

  &:hover {
    @extend %hoverStyle;
  }

  &Selected {
    &:hover {
      @extend %hoverStyle;
    }
  }
}

%hoverStyle {
  background-color: $skyBlue100 !important;
  color: $white !important;
}
