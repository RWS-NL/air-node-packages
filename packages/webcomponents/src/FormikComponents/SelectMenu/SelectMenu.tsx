import { FormControl, FormHelperText, InputLabel, MenuItem, Select } from '@material-ui/core';
import ArrowDropIcon from '@material-ui/icons/ArrowDropDown';
import { objectHasProperty } from '@rws-air/utils';
import classnames from 'classnames';
import { fieldToSelect, SelectProps } from 'formik-material-ui';
import React, { FC, Fragment, useEffect, useRef, useState } from 'react';
import { When } from 'react-if';
import css from './SelectMenu.scss';

export interface SelectMenuProps extends SelectProps {
  /** The label of the select menu */
  label: string;
  /** The selectable options for the select menu */
  options: { value: string | number | string[]; label: string }[];
}

/**
 * Constructs a Select Menu with the formik validation
 * @param props Props to pass to the select menu component
 * @example
 * <FastField
 *   name='type' type='text' label='Example'
 *   options={[ { label: 'john', value='john'}, { label: 'connor', value: 'connor'} ]}
 *   placeholder='Example placeholder' variant='outlined'
 *   data-qa='exampleDataQA' component={SelectMenu} required
 * />
 */
export const SelectMenu: FC<SelectMenuProps> = ({ options, ...props }) => {
  const inputLabel = useRef<HTMLLabelElement>(null);
  const [ labelWidth, setLabelWidth ] = useState(0);
  useEffect(() => {
    setLabelWidth(inputLabel.current!.offsetWidth);
  }, []);

  const selectFieldHasErrors = objectHasProperty(props.form.errors, props.field.name) && objectHasProperty(props.form.touched, props.field.name);

  return (
    <Fragment>
      <FormControl variant='outlined' fullWidth>
        <InputLabel
          classes={{ root: css.inputLabel, shrink: css.inputLabelShrink }}
          ref={inputLabel} id={`validated-select-menu-${props.label}`}
          required={objectHasProperty(props, 'required') ? props.required : false}
        >
          {props.label}
        </InputLabel>
        <Select
          {...fieldToSelect(props)}
          type='text'
          labelId={`validated-select-menu-${props.label}`}
          labelWidth={labelWidth}
          className={classnames([ css.select, { [css.errorSelect]: selectFieldHasErrors } ])}
          IconComponent={() => <ArrowDropIcon color={selectFieldHasErrors ? 'error' : 'primary'} />}
          fullWidth
          displayEmpty
          value={props.field.value || ''}
          onChange={props.field.onChange}
          onBlur={props.field.onBlur}
          autoFocus={objectHasProperty(props, 'autoFocus') ? props.autoFocus : false}
          required={objectHasProperty(props, 'required') ? props.required : false}
          inputProps={{ classes: { root: css.input } }}
        >
          {options.map((item, index) => (
            <MenuItem classes={{ root: css.menuItem, selected: css.menuItemSelected }} key={index + 1} value={item.value}>{item.label}</MenuItem>
          ))}
        </Select>
        <When condition={selectFieldHasErrors}>
          <FormHelperText component='div' classes={{ root: css.errorLabel }}>{props.form.errors[props.field.name]}</FormHelperText>
        </When>
      </FormControl>
    </Fragment>
  );
};

export default SelectMenu;