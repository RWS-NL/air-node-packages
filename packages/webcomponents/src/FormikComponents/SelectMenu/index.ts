export { default as SelectMenu, SelectMenuProps } from './SelectMenu';