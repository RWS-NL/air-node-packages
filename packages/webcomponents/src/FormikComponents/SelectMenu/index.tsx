import { FormControl, FormHelperText, InputLabel, MenuItem } from '@material-ui/core';
import ArrowDropIcon from '@material-ui/icons/ArrowDropDown';
import clsx from 'clsx';
import { useField } from 'formik';
import { fieldToSelect, Select, SelectProps } from 'formik-material-ui';
import React, { Fragment, useEffect, useRef, useState } from 'react';
import { When } from 'react-if';
import css from './SelectMenu.scss';

/** The possible types for the option value */
export type SelectMenuOptionValues = string | number | string[];

export interface SelectMenuOption<T extends SelectMenuOptionValues> {
  /** The value for the option */
  value: T;
  /** The label shown next to the radio button */
  label: string;
}

export interface SelectMenuProps<T extends SelectMenuOptionValues> extends SelectProps {
  /** The label of the select menu */
  label: string;
  /** The selectable options for the select menu */
  options: SelectMenuOption<T>[];
}

/**
 * Constructs a Select Menu with the formik validation
 * @param props Props to pass to the select menu component
 * @example
 * ```jsx
 * <Field
 *   component={SelectMenu}
 *   name='type'
 *   type='text'
 *   required
 *   placeholder='Example Placeholder'
 *   variant='outlined'
 *   data-qa='sample-select-menu'
 *   label='Example'
 *   options={ [{ value: 'JOHN', label: 'John'}, { value: 'CONNOR', label: 'Connor'}] }
 * />
 * ```
 */
export const SelectMenu = <T extends SelectMenuOptionValues>({
  label,
  options,
  placeholder,
  className,
  autoFocus,
  required,
  ...props
}: SelectMenuProps<T>) => {
  const [{ onChange, onBlur }, { error, touched }] = useField(props.field.name);

  const inputLabel = useRef<HTMLLabelElement>(null);
  const [labelWidth, setLabelWidth] = useState(0);

  useEffect(() => {
    if (inputLabel.current) {
      setLabelWidth(inputLabel.current.offsetWidth);
    }
  }, []);

  const selectFieldHasErrors = Boolean(error) && touched;

  return (
    <Fragment>
      <FormControl variant='outlined' fullWidth>
        <InputLabel
          classes={{ root: css.inputLabel, shrink: css.inputLabelShrink }}
          ref={inputLabel}
          id={`validated-select-menu-${label}`}
          required={required}
        >
          {label}
        </InputLabel>
        <Select
          {...props}
          {...fieldToSelect(props)}
          type='text'
          labelId={`validated-select-menu-${label}`}
          labelWidth={labelWidth}
          className={clsx([css.select, { [css.errorSelect]: selectFieldHasErrors }])}
          IconComponent={() => <ArrowDropIcon color={selectFieldHasErrors ? 'error' : 'primary'} />}
          fullWidth
          displayEmpty
          onChange={onChange}
          onBlur={onBlur}
          autoFocus={autoFocus}
          required={required}
          inputProps={{ classes: { root: css.input } }}
          MenuProps={{
            getContentAnchorEl: null,
            anchorOrigin: {
              vertical: 'bottom',
              horizontal: 'center'
            },
            transformOrigin: {
              vertical: 'top',
              horizontal: 'center'
            },
            elevation: 2,
            classes: {
              list: css.selectEntry
            }
          }}
        >
          {options.map((item, index) => (
            <MenuItem
              classes={{ root: css.menuItem, selected: css.menuItemSelected }}
              key={index + 1}
              value={item.value}
            >
              {item.label}
            </MenuItem>
          ))}
        </Select>
        <When condition={selectFieldHasErrors}>
          <FormHelperText component='div' classes={{ root: css.errorLabel }}>
            {error || ''}
          </FormHelperText>
        </When>
      </FormControl>
    </Fragment>
  );
};
