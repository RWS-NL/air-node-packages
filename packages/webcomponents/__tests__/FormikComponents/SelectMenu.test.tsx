import { shallow } from 'enzyme';
import React from 'react';
import SelectMenu from '../../src/FormikComponents/SelectMenu/SelectMenu';

describe('Snapshot Testing', () => {
  test('Required Props', () => {
    const fieldMock = { name: '' };
    const mockForm = { errors: [] };

    const selectMenu = shallow(
      //@ts-ignore
      <SelectMenu label='label' options={[ { value: 'one', label: 'one' }, { value: 'two', label: 'two' } ]} type='text' field={fieldMock} form={mockForm} />
    );
    expect(selectMenu).toMatchSnapshot();
  });
});