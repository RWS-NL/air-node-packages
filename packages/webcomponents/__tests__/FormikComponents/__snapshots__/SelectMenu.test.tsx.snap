// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing Required Props 1`] = `
<Fragment>
  <WithStyles(ForwardRef(FormControl))
    fullWidth={true}
    variant="outlined"
  >
    <WithStyles(ForwardRef(InputLabel))
      classes={
        Object {
          "root": "inputLabel",
          "shrink": "inputLabelShrink",
        }
      }
      id="validated-select-menu-label"
      required={false}
    >
      label
    </WithStyles(ForwardRef(InputLabel))>
    <WithStyles(ForwardRef(Select))
      IconComponent={[Function]}
      autoFocus={false}
      className="select"
      displayEmpty={true}
      fullWidth={true}
      inputProps={
        Object {
          "classes": Object {
            "root": "input",
          },
        }
      }
      label="label"
      labelId="validated-select-menu-label"
      labelWidth={0}
      name=""
      required={false}
      type="text"
      value=""
    >
      <WithStyles(ForwardRef(MenuItem))
        classes={
          Object {
            "root": "menuItem",
            "selected": "menuItemSelected",
          }
        }
        key="1"
        value="one"
      >
        one
      </WithStyles(ForwardRef(MenuItem))>
      <WithStyles(ForwardRef(MenuItem))
        classes={
          Object {
            "root": "menuItem",
            "selected": "menuItemSelected",
          }
        }
        key="2"
        value="two"
      >
        two
      </WithStyles(ForwardRef(MenuItem))>
    </WithStyles(ForwardRef(Select))>
    <When
      condition={false}
    >
      <WithStyles(ForwardRef(FormHelperText))
        classes={
          Object {
            "root": "errorLabel",
          }
        }
        component="div"
      />
    </When>
  </WithStyles(ForwardRef(FormControl))>
</Fragment>
`;
